File: internal/capability/capability.go
// SPDX-License-Identifier: Apache-2.0

// Package capability is the Observation-and-Control Contract registry described
// in ADR-0001.
//
// Every ability VayuCell might have on a device is a registered Capability whose
// obligations are expressed as types with INVALID ZERO VALUES. A capability that
// leaves any obligation unanswered fails a test rather than a review, so nothing
// lands undeclared.
//
// Two rules here are load-bearing and are enforced by Complete rather than by
// convention:
//
//   - Verify may never be nil. A control that cannot be read back after being
//     set is indistinguishable from one that silently stopped working, and
//     reporting it would be the exact lie CHARTER Article IV forbids.
//
//   - A safety-class capability may not degrade quietly. Where a safety control
//     is absent it must refuse the operation, so the panel can render a
//     permanent failing row instead of a soft warning nobody reads.
package capability

import (
	"context"
	"errors"
	"fmt"
	"sort"
	"strings"
)

// Tier is the environment a device provides, per ADR-0001 §1.
//
// A tier is DETECTED from positive evidence, never inferred from a model name
// or a version number. TierUnset is not a valid answer: a device whose tier
// could not be established is refused, not defaulted.
type Tier uint8

const (
	TierUnset Tier = iota // zero value — invalid, never a result
	T0                    // stock, unprivileged userspace
	T1                    // stock, root available
	T2                    // virtualised Linux guest
	T3                    // mainline Linux
)

var tierNames = map[Tier]string{T0: "T0", T1: "T1", T2: "T2", T3: "T3"}

func (t Tier) String() string {
	if n, ok := tierNames[t]; ok {
		return n
	}
	return "tierUnset"
}

// Valid reports whether t is a real tier rather than the zero value.
func (t Tier) Valid() bool { _, ok := tierNames[t]; return ok }

// AtLeast reports whether t satisfies a capability whose floor is min.
// An invalid tier satisfies nothing.
func (t Tier) AtLeast(min Tier) bool {
	if !t.Valid() || !min.Valid() {
		return false
	}
	return t >= min
}

// Class is what kind of thing a capability governs. It decides how strictly the
// registry treats absence.
type Class uint8

const (
	ClassUnset   Class = iota // zero value — invalid
	ClassSafety               // battery, thermal — never degrades quietly
	ClassServing              // ingress, site serving
	ClassStorage              // durability, wear
	ClassNetwork              // reachability, egress
	ClassObserve              // read-only measurement; may have no Apply
)

var classNames = map[Class]string{
	ClassSafety: "safety", ClassServing: "serving",
	ClassStorage: "storage", ClassNetwork: "network", ClassObserve: "observe",
}

func (c Class) String() string {
	if n, ok := classNames[c]; ok {
		return n
	}
	return "classUnset"
}

func (c Class) Valid() bool { _, ok := classNames[c]; return ok }

// Disposition is what happens when a capability is absent on this device.
type Disposition uint8

const (
	DispUnset   Disposition = iota // zero value — invalid
	DispDegrade                    // continue with reduced function
	DispRefuse                     // refuse the dependent operation
)

var dispNames = map[Disposition]string{DispDegrade: "degrade", DispRefuse: "refuse"}

func (d Disposition) String() string {
	if n, ok := dispNames[d]; ok {
		return n
	}
	return "dispUnset"
}

func (d Disposition) Valid() bool { _, ok := dispNames[d]; return ok }

// Result is the outcome of detecting or verifying a capability.
//
// There is deliberately no "ok" or "pass" member. Present means the mechanism
// exists; whether it is WORKING is a separate question answered by Verify.
// Unverified is never treated as Absent, and Absent is never treated as
// protection (CHARTER Article IV).
type Result uint8

const (
	ResultUnset Result = iota // zero value — invalid, never reported
	Present                   // mechanism found on this device
	Absent                    // mechanism genuinely not present
	Unverified                // could not be established — NOT the same as Absent
)

var resultNames = map[Result]string{Present: "present", Absent: "absent", Unverified: "unverified"}

func (r Result) String() string {
	if n, ok := resultNames[r]; ok {
		return n
	}
	return "resultUnset"
}

func (r Result) Valid() bool { _, ok := resultNames[r]; return ok }

// Observation is a Result with the evidence that produced it. The evidence is
// mandatory: a bare verdict with no reason cannot be audited, and an operator
// reading "absent" deserves to know what was looked for.
type Observation struct {
	Result   Result
	Evidence string // e.g. the node path that answered, or why nothing did
}

// ID identifies a capability. Stable across releases; it appears in the
// hardware database and in operator-facing reports.
type ID string

// DetectFn establishes whether the mechanism exists on this device.
type DetectFn func(ctx context.Context) (Observation, error)

// ApplyFn sets the capability to a target value. May be nil ONLY for
// ClassObserve capabilities, which measure rather than control.
type ApplyFn func(ctx context.Context, target string) error

// VerifyFn reads the state back from the hardware or kernel.
//
// This is the whole point of the package. It may never be nil.
type VerifyFn func(ctx context.Context) (Observation, error)

// Capability is one contract. Every field is an obligation.
type Capability struct {
	ID        ID
	Floor     Tier        // lowest tier that can provide this
	Class     Class       // what kind of thing this governs
	Detect    DetectFn    // is the mechanism here?
	Apply     ApplyFn     // set it (nil only when Class == ClassObserve)
	Verify    VerifyFn    // read it back — NEVER nil
	OnAbsent  Disposition // degrade, or refuse?
	Rationale string      // why these answers, in prose, shown to the operator
}

// Errors returned by Complete. Exported so tests can assert on the reason
// rather than on a string.
var (
	ErrNoID           = errors.New("capability: ID is empty")
	ErrFloorUnset     = errors.New("capability: Floor is tierUnset")
	ErrClassUnset     = errors.New("capability: Class is classUnset")
	ErrDispUnset      = errors.New("capability: OnAbsent is dispUnset")
	ErrNoDetect       = errors.New("capability: Detect is nil")
	ErrNoVerify       = errors.New("capability: Verify is nil — a control that cannot be read back may not be reported (CHARTER Art. IV)")
	ErrNoApply        = errors.New("capability: Apply is nil on a controlling capability (only ClassObserve may omit it)")
	ErrNoRationale    = errors.New("capability: Rationale is empty")
	ErrSafetyDegrades = errors.New("capability: a ClassSafety capability may not use DispDegrade — safety controls refuse or are reported permanently failing (CHARTER Art. III)")
)

// Complete reports whether every obligation was answered.
//
// It is called for every capability by a registry test, so a capability added
// without deciding its policy fails the build.
func (c Capability) Complete() error {
	var errs []error
	if strings.TrimSpace(string(c.ID)) == "" {
		errs = append(errs, ErrNoID)
	}
	if !c.Floor.Valid() {
		errs = append(errs, ErrFloorUnset)
	}
	if !c.Class.Valid() {
		errs = append(errs, ErrClassUnset)
	}
	if !c.OnAbsent.Valid() {
		errs = append(errs, ErrDispUnset)
	}
	if c.Detect == nil {
		errs = append(errs, ErrNoDetect)
	}
	if c.Verify == nil {
		errs = append(errs, ErrNoVerify)
	}
	if c.Apply == nil && c.Class != ClassObserve {
		errs = append(errs, ErrNoApply)
	}
	if strings.TrimSpace(c.Rationale) == "" {
		errs = append(errs, ErrNoRationale)
	}
	if c.Class == ClassSafety && c.OnAbsent == DispDegrade {
		errs = append(errs, ErrSafetyDegrades)
	}
	if len(errs) == 0 {
		return nil
	}
	return fmt.Errorf("capability %q: %w", c.ID, errors.Join(errs...))
}

// Registry holds every declared capability.
//
// Registration is fallible on purpose: an incomplete capability is refused at
// the point of registration rather than discovered at runtime.
type Registry struct {
	caps map[ID]Capability
}

// NewRegistry returns an empty registry.
func NewRegistry() *Registry { return &Registry{caps: make(map[ID]Capability)} }

// ErrDuplicate is returned when an ID is registered twice.
var ErrDuplicate = errors.New("capability: already registered")

// Register validates and adds a capability.
func (r *Registry) Register(c Capability) error {
	if err := c.Complete(); err != nil {
		return err
	}
	if _, dup := r.caps[c.ID]; dup {
		return fmt.Errorf("%w: %q", ErrDuplicate, c.ID)
	}
	r.caps[c.ID] = c
	return nil
}

// MustRegister panics if the capability is incomplete. Intended for package
// initialisation of compiled-in capabilities, where a failure is a build-time
// programming error rather than a runtime condition.
func (r *Registry) MustRegister(c Capability) {
	if err := r.Register(c); err != nil {
		panic(err)
	}
}

// Get returns a capability by ID.
func (r *Registry) Get(id ID) (Capability, bool) { c, ok := r.caps[id]; return c, ok }

// Len reports how many capabilities are registered.
func (r *Registry) Len() int { return len(r.caps) }

// IDs returns every registered ID, sorted, so reports are deterministic.
func (r *Registry) IDs() []ID {
	out := make([]ID, 0, len(r.caps))
	for id := range r.caps {
		out = append(out, id)
	}
	sort.Slice(out, func(i, j int) bool { return out[i] < out[j] })
	return out
}

// Validate re-checks every registered capability. Called by the registry test
// so that a capability mutated after registration still fails the build.
func (r *Registry) Validate() error {
	var errs []error
	for _, id := range r.IDs() {
		if err := r.caps[id].Complete(); err != nil {
			errs = append(errs, err)
		}
	}
	return errors.Join(errs...)
}

File: core/src/capability.rs
// SPDX-License-Identifier: Apache-2.0

//! The Observation-and-Control Contract registry described in ADR-0001.
//!
//! Every ability VayuCell might have on a device is a [`Capability`] whose
//! obligations are expressed as types. ADR-0005 §2.1 is the reason this is
//! Rust: two of the rules that needed runtime guards in the first
//! implementation are now facts about the type system.
//!
//! * [`Capability::verify`] is not an [`Option`]. **A capability with no
//!   read-back does not compile.** That is the single most important rule in
//!   the project — a control that cannot be read back after being set is
//!   indistinguishable from one that silently stopped working, and reporting it
//!   would be CHARTER Article IV violated by the reporting mechanism itself.
//!
//! * [`Tier`] has no `Unset` variant. An undetected tier is `None`, which
//!   cannot be compared against a floor at all, so there is no zero value to
//!   defend against.
//!
//! Two rules genuinely need checking and are enforced by [`Capability::check`]:
//! a safety-class capability may not degrade quietly, and only an observe-class
//! capability may omit `apply`.
//!
//! # What cannot be written down
//!
//! These are proofs, not tests: they are compiled by `cargo test` and fail the
//! build if they ever start compiling.
//!
//! A capability with no read-back does not exist:
//!
//! ```compile_fail
//! use vayucell_core::capability::{Capability, Class, Disposition, Observation, ProbeError, Tier};
//! fn d() -> Result<Observation, ProbeError> { unimplemented!() }
//! let _ = Capability {
//!     id: "x", floor: Tier::T0, class: Class::Observe,
//!     detect: d, apply: None,
//!     on_absent: Disposition::Refuse, rationale: "x",
//! };  // error: missing field `verify`
//! ```
//!
//! A read-back cannot be declared absent, which is the invariant itself
//! rather than the generic "no field may be omitted" rule:
//!
//! ```compile_fail
//! use vayucell_core::capability::{Capability, Class, Disposition, Observation, ProbeError, Result_, Tier};
//! fn d() -> Result<Observation, ProbeError> { Ok(Observation::new(Result_::Present, "e")) }
//! let _ = Capability {
//!     id: "x", floor: Tier::T0, class: Class::Observe,
//!     detect: d, apply: None,
//!     verify: None,  // error: expected fn pointer, found `Option<_>`
//!     on_absent: Disposition::Refuse, rationale: "x",
//! };
//! ```
//!
//! There is no unset tier to reach for:
//!
//! ```compile_fail
//! use vayucell_core::capability::Tier;
//! let _t = Tier::Unset;  // error: no variant named `Unset`
//! ```
//!
//! And for contrast, the same capability WITH a read-back compiles:
//!
//! ```
//! use vayucell_core::capability::{Capability, Class, Disposition, Observation, ProbeError, Result_, Tier};
//! fn d() -> Result<Observation, ProbeError> { Ok(Observation::new(Result_::Present, "e")) }
//! let c = Capability {
//!     id: "x", floor: Tier::T0, class: Class::Observe,
//!     detect: d, apply: None, verify: d,
//!     on_absent: Disposition::Refuse, rationale: "x",
//! };
//! assert!(c.check().is_ok());
//! ```

use std::collections::BTreeMap;
use std::fmt;

/// The environment a device provides, per ADR-0001 §1.
///
/// A tier is **detected from positive evidence**, never inferred from a model
/// name or a version number. There is deliberately no `Unset` variant: a device
/// whose tier could not be established is represented as `None` and satisfies
/// nothing.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Tier {
    /// Stock Android, unprivileged userspace.
    T0,
    /// Stock Android with root available.
    T1,
    /// Virtualised Linux guest (protected KVM).
    T2,
    /// Mainline Linux (postmarketOS-class).
    T3,
}

impl Tier {
    /// Reports whether this tier satisfies a capability whose floor is `min`.
    #[must_use]
    pub fn at_least(self, min: Tier) -> bool {
        self >= min
    }
}

impl fmt::Display for Tier {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            Tier::T0 => "T0",
            Tier::T1 => "T1",
            Tier::T2 => "T2",
            Tier::T3 => "T3",
        };
        f.write_str(s)
    }
}

/// Whether a detected tier satisfies a floor.
///
/// An undetected tier (`None`) satisfies nothing. This is a free function
/// rather than a method so the absent case cannot be forgotten at a call site.
#[must_use]
pub fn tier_satisfies(detected: Option<Tier>, floor: Tier) -> bool {
    matches!(detected, Some(t) if t.at_least(floor))
}

/// What kind of thing a capability governs. Decides how strictly absence is
/// treated.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Class {
    /// Battery and thermal controls. Never degrades quietly (CHARTER Art. III).
    Safety,
    /// Ingress and site serving.
    Serving,
    /// Durability and wear.
    Storage,
    /// Reachability and egress.
    Network,
    /// Read-only measurement. The only class that may omit `apply`.
    Observe,
}

impl fmt::Display for Class {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            Class::Safety => "safety",
            Class::Serving => "serving",
            Class::Storage => "storage",
            Class::Network => "network",
            Class::Observe => "observe",
        };
        f.write_str(s)
    }
}

/// What happens when a capability is absent on this device.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Disposition {
    /// Continue with reduced function.
    Degrade,
    /// Refuse the dependent operation.
    Refuse,
}

/// The outcome of detecting or verifying a capability.
///
/// There is deliberately no `Ok` or `Pass` member. [`Result_::Present`] means
/// the mechanism exists; whether it *works* is a separate question answered by
/// verification. [`Result_::Unverified`] is never treated as
/// [`Result_::Absent`], and absence is never treated as protection
/// (CHARTER Article IV).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[allow(clippy::module_name_repetitions)]
pub enum Result_ {
    /// The mechanism was found on this device.
    Present,
    /// The mechanism is genuinely not present.
    Absent,
    /// Could not be established. **Not** the same as `Absent`.
    Unverified,
}

impl fmt::Display for Result_ {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            Result_::Present => "present",
            Result_::Absent => "absent",
            Result_::Unverified => "unverified",
        };
        f.write_str(s)
    }
}

/// A [`Result_`] with the evidence that produced it.
///
/// The evidence is mandatory: a bare verdict cannot be audited, and an operator
/// reading "absent" deserves to know what was looked for.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Observation {
    /// What was concluded.
    pub result: Result_,
    /// Why — the node that answered, or what was looked for and not found.
    pub evidence: String,
}

impl Observation {
    /// Builds an observation.
    #[must_use]
    pub fn new(result: Result_, evidence: impl Into<String>) -> Self {
        Self {
            result,
            evidence: evidence.into(),
        }
    }
}

/// Errors a probe may return. Distinct from an [`Observation`]: a failure to
/// *run* the probe is not a finding about the device.
///
/// The string is the reason, for the operator.
#[derive(Debug)]
pub struct ProbeError(pub String);

impl fmt::Display for ProbeError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "probe failed: {}", self.0)
    }
}

impl std::error::Error for ProbeError {}

/// Establishes whether the mechanism exists on this device.
pub type DetectFn = fn() -> Result<Observation, ProbeError>;

/// Sets the capability to a target value.
pub type ApplyFn = fn(&str) -> Result<(), ProbeError>;

/// Reads the state back from the hardware or kernel.
///
/// Never optional. See the module documentation.
pub type VerifyFn = fn() -> Result<Observation, ProbeError>;

/// One contract. Every field is an obligation, and the struct cannot be built
/// without answering all of them.
#[derive(Debug, Clone)]
pub struct Capability {
    /// Stable identifier, used in reports and the hardware database.
    pub id: &'static str,
    /// Lowest tier that can provide this.
    pub floor: Tier,
    /// What kind of thing this governs.
    pub class: Class,
    /// Is the mechanism here?
    pub detect: DetectFn,
    /// Set it. `None` is permitted **only** for [`Class::Observe`].
    pub apply: Option<ApplyFn>,
    /// Read it back. Not an `Option` — see the module documentation.
    pub verify: VerifyFn,
    /// Degrade, or refuse?
    pub on_absent: Disposition,
    /// Why these answers, in prose, shown to the operator.
    pub rationale: &'static str,
}

/// Why a capability was refused.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Invalid {
    /// The identifier was empty or blank.
    NoId,
    /// The rationale was empty or blank.
    NoRationale,
    /// A controlling capability omitted `apply`.
    NoApply,
    /// A safety capability tried to degrade quietly.
    SafetyDegrades,
    /// The identifier was already registered.
    Duplicate,
}

impl fmt::Display for Invalid {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            Invalid::NoId => "id is empty",
            Invalid::NoRationale => "rationale is empty",
            Invalid::NoApply => {
                "apply is None on a controlling capability (only Class::Observe may omit it)"
            }
            Invalid::SafetyDegrades => {
                "a Class::Safety capability may not use Disposition::Degrade — safety controls \
                 refuse or are reported permanently failing (CHARTER Art. III)"
            }
            Invalid::Duplicate => "already registered",
        };
        f.write_str(s)
    }
}

impl std::error::Error for Invalid {}

impl Capability {
    /// Checks the obligations the compiler cannot.
    ///
    /// The two rules that *did* need runtime guards in the first implementation
    /// — a missing `verify`, and a tier zero value — are gone: the first does
    /// not compile, and the second does not exist.
    ///
    /// # Errors
    /// Returns every way in which the capability is incomplete.
    pub fn check(&self) -> Result<(), Vec<Invalid>> {
        let mut errs = Vec::new();
        if self.id.trim().is_empty() {
            errs.push(Invalid::NoId);
        }
        if self.rationale.trim().is_empty() {
            errs.push(Invalid::NoRationale);
        }
        if self.apply.is_none() && self.class != Class::Observe {
            errs.push(Invalid::NoApply);
        }
        if self.class == Class::Safety && self.on_absent == Disposition::Degrade {
            errs.push(Invalid::SafetyDegrades);
        }
        if errs.is_empty() {
            Ok(())
        } else {
            Err(errs)
        }
    }
}

/// Every declared capability.
///
/// A `BTreeMap` so iteration is ordered and reports are deterministic without
/// an explicit sort.
#[derive(Debug, Default)]
pub struct Registry {
    caps: BTreeMap<&'static str, Capability>,
}

impl Registry {
    /// An empty registry.
    #[must_use]
    pub fn new() -> Self {
        Self {
            caps: BTreeMap::new(),
        }
    }

    /// Validates and adds a capability.
    ///
    /// # Errors
    /// Returns the reasons the capability was refused.
    pub fn register(&mut self, c: Capability) -> Result<(), Vec<Invalid>> {
        c.check()?;
        if self.caps.contains_key(c.id) {
            return Err(vec![Invalid::Duplicate]);
        }
        self.caps.insert(c.id, c);
        Ok(())
    }

    /// Looks a capability up.
    #[must_use]
    pub fn get(&self, id: &str) -> Option<&Capability> {
        self.caps.get(id)
    }

    /// How many capabilities are registered.
    #[must_use]
    pub fn len(&self) -> usize {
        self.caps.len()
    }

    /// Whether the registry is empty.
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.caps.is_empty()
    }

    /// Every registered identifier, in sorted order.
    #[must_use]
    pub fn ids(&self) -> Vec<&'static str> {
        self.caps.keys().copied().collect()
    }

    /// Re-checks every registered capability, so one mutated after
    /// registration still fails the build.
    ///
    /// # Errors
    /// Returns each offending identifier with its reasons.
    pub fn validate(&self) -> Result<(), Vec<(&'static str, Vec<Invalid>)>> {
        let mut errs = Vec::new();
        for (id, c) in &self.caps {
            if let Err(e) = c.check() {
                errs.push((*id, e));
            }
        }
        if errs.is_empty() {
            Ok(())
        } else {
            Err(errs)
        }
    }

    /// Capabilities this device can provide, given its detected tier.
    ///
    /// An undetected tier yields nothing, per [`tier_satisfies`].
    #[must_use]
    pub fn available_at(&self, detected: Option<Tier>) -> Vec<&'static str> {
        self.caps
            .values()
            .filter(|c| tier_satisfies(detected, c.floor))
            .map(|c| c.id)
            .collect()
    }
}

#[cfg(test)]
#[allow(clippy::unnecessary_wraps)]
mod internal_tests {
    use super::*;

    fn d() -> Result<Observation, ProbeError> {
        Ok(Observation::new(Result_::Present, "e"))
    }

    #[test]
    fn validate_catches_a_capability_mutated_after_registration() {
        // register() refuses a degrading safety capability, so the only way to
        // get one into the map is to bypass registration -- which a careless
        // refactor could do. validate() is the backstop, and this test has the
        // module-private access needed to prove it works.
        let mut r = Registry::new();
        r.caps.insert(
            "forced.in",
            Capability {
                id: "forced.in",
                floor: Tier::T0,
                class: Class::Safety,
                detect: d,
                apply: None,
                verify: d,
                on_absent: Disposition::Degrade, // never survives register()
                rationale: "inserted without validation",
            },
        );

        let errs = r
            .validate()
            .expect_err("validate must catch a bypassed capability");
        assert_eq!(errs.len(), 1);
        assert_eq!(errs[0].0, "forced.in");
        assert!(
            errs[0].1.contains(&Invalid::SafetyDegrades),
            "got {:?}",
            errs[0].1
        );
    }
}
